models/src/main/scala/coop/rchain/models/PrettyPrinter.scala

```diff
@@ -9,220 +9,170 @@ import coop.rchain.models.Var.VarInstance
+
+      case e: Eval => "*" + buildString(e.channel.get)
       case n: New =>
-        print("new ")
-        printNewVariables(n.bindCount)
-        print(" in {")
-        prettyPrint(n.p.get)
-        print("}")
+        "new" + buildNewVariables(n.bindCount) + " in { " + PrettyPrinter(
+          freeShift,
+          boundShift + n.bindCount).buildString(n.p.get) + " }"
       case e: Expr =>
-        prettyPrint(e)
+        buildString(e)
       case m: Match =>
-        print("match { ")
-        prettyPrint(m.target.get)
-        print(" } { ")
-        for { (matchCase, i) <- m.cases.zipWithIndex } yield {
-          prettyPrint(matchCase.pattern.get)
-          print(" => ")
-          prettyPrint(matchCase.source.get)
-          if (i != m.cases.length - 1) {
-            print("; ")
+        "match { " + buildString(m.target.get) + " } { " + {
+          for { (matchCase, i) <- m.cases.zipWithIndex } yield {
+            buildString(matchCase.pattern.get) + " => " + buildString(matchCase.source.get) + {
+              if (i != m.cases.length - 1) " ; "
+              else ""
+            }
           }
-        }
-        print(" }")
-      case g: GPrivate =>
-        print(g.id)
+        } + " }"
+      case g: GPrivate => g.id
       case p: Par =>
-        if (isEmpty(p)) {
-          print("Nil")
-        } else {
+        if (isEmpty(p)) "Nil"
+        else {
           val list = List(p.sends, p.receives, p.evals, p.news, p.exprs, p.matches, p.ids)
-          list.foldLeft(false)((acc: Boolean, items: Seq[GeneratedMessage]) => {
-            if (items.nonEmpty) {
-              if (acc)
-                print(" | ")
-              for { (item, i) <- items.zipWithIndex } yield {
-                prettyPrint(item)
-                if (i != items.length - 1) {
-                  print(" | ")
-                }
-              }
-              true
-            } else {
-              acc
-            }
-          })
-        }
+          ((false, "") /: list) {
+            case ((prevNonEmpty, string), items) =>
+              if (items.nonEmpty) {
+                (true, string + { if (prevNonEmpty) " | " else "" } + ("" /: items.zipWithIndex) {
+                  case (_string, (_p, i)) =>
+                    _string + buildString(_p) + {
+                      if (i != items.length - 1) " | " else ""
+                    }
+                })
+              } else (prevNonEmpty, string)
+          }
+        }._2
       case _ => throw new Error("Attempt to print unknown GeneratedMessage type")
     }
 
   // TODO: Calculate proper numbering/naming
-  private def printNewVariables(bindCount: Int) = {
-    // We arbitrarily limit the new variable printing count to MAX_NEW_VAR_COUNT
+  def buildNewVariables(bindCount: Int): String = {
+    // We arbitrarily limit the new variable string count to MAX_NEW_VAR_COUNT
     // to prevent exploding the state of the shapeless generator
     val MAX_NEW_VAR_COUNT = 128
```

Comment:
Above the limit, we should just print xk, xk+1, ... xk + count - 1

Reply:
I think I understand. I need to do 

`buildSeq((0 until Math.min(MAX_NEW_VAR_COUNT, bindCount)).map(i =>
   Expr(exprInstance = GString(s" x${boundShift + i}"))))`